In [ ]:
import pandas as pd
import numpy as np

In [ ]:
# data = pd.read_csv("/content/drive/MyDrive/Data Science Project/Datasets/preprocessing_stage2.csv")
data=pd.read_csv("preprocessing_stage2.csv")

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 56 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1460 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   LotShape       1460 non-null   object 
 6   LandContour    1460 non-null   object 
 7   LotConfig      1460 non-null   object 
 8   LandSlope      1460 non-null   object 
 9   Neighborhood   1460 non-null   object 
 10  BldgType       1460 non-null   object 
 11  HouseStyle     1460 non-null   object 
 12  OverallQual    1460 non-null   int64  
 13  OverallCond    1460 non-null   int64  
 14  YearBuilt      1460 non-null   int64  
 15  YearRemodAdd   1460 non-null   int64  
 16  RoofStyle      1460 non-null   object 
 17  Exterior1st    1460 non-null   object 
 18  Exterior

In [ ]:
X_train = data.drop(['SalePrice', 'Unnamed: 0'], axis=1)
y_train = data['SalePrice']

In [ ]:
numerical_cols = X_train.select_dtypes(include=np.number).columns.tolist()
categorical_cols = X_train.select_dtypes(include='object').columns.tolist()

print("Numerical Columns:", numerical_cols)
print("Categorical Columns:", categorical_cols)

Numerical Columns: ['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'GrLivArea', 'BsmtFullBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch']
Categorical Columns: ['MSZoning', 'LotShape', 'LandContour', 'LotConfig', 'LandSlope', 'Neighborhood', 'BldgType', 'HouseStyle', 'RoofStyle', 'Exterior1st', 'Exterior2nd', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual', 'GarageType', 'GarageFinish', 'PavedDrive', 'SaleType', 'SaleCondition']


In [ ]:
numerical_features_standard = ['BsmtFinSF1', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'GarageYrBlt', 'OpenPorchSF']
numerical_features_robust = ["LotFrontage", "LotArea", "MasVnrArea", "GrLivArea", "GarageArea", "WoodDeckSF", "OpenPorchSF", "EnclosedPorch"]
numerical_features_minmax = [col for col in numerical_cols if col not in numerical_features_standard and col not in numerical_features_robust]

print("Numerical features to be scaled with StandardScaler:", numerical_features_standard)
print("Numerical features to be scaled with MinMaxScaler:", numerical_features_minmax)
print("Numerical features to be scaled with RobustScaler:", numerical_features_robust)

Numerical features to be scaled with StandardScaler: ['BsmtFinSF1', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'GarageYrBlt', 'OpenPorchSF']
Numerical features to be scaled with MinMaxScaler: ['MSSubClass', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'BsmtFullBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageCars']
Numerical features to be scaled with RobustScaler: ['LotFrontage', 'LotArea', 'MasVnrArea', 'GrLivArea', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch']


In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler, OneHotEncoder

transformers = [
    ('scaler_standard', StandardScaler(), numerical_features_standard),
    ('scaler_minmax', MinMaxScaler(), numerical_features_minmax),
    ('scaler_robust', RobustScaler(), numerical_features_robust),
    ('onehot', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
]

preprocessor = ColumnTransformer(transformers=transformers)

Train your model here


In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor

# Create a Random Forest Regressor model
rf_regressor = RandomForestRegressor(n_estimators=100, random_state=42) # You can adjust n_estimators and random_state


In [ ]:
# X_test = pd.read_csv("/content/drive/MyDrive/Data Science Project/Datasets/preprocessed_test_data (1).csv")
X_test=pd.read_csv("preprocessed_test_data (1).csv")

In [ ]:
X_test = X_test.drop(['Id'], axis=1)


In [ ]:
X_test.shape

(1459, 54)

In [ ]:
X_train.head(2)

MSSubClass MSZoning  LotFrontage  LotArea LotShape LandContour LotConfig  \
0          60       RL         65.0     8450      Reg         Lvl    Inside   
1          20       RL         80.0     9600      Reg         Lvl       FR2   

  LandSlope Neighborhood BldgType  ... GarageYrBlt  GarageFinish  GarageCars  \
0       Gtl      CollgCr     1Fam  ...      2003.0           RFn           2   
1       Gtl      Veenker     1Fam  ...      1976.0           RFn           2   

   GarageArea  PavedDrive WoodDeckSF OpenPorchSF EnclosedPorch  SaleType  \
0         548           Y          0          61             0        WD   
1         460           Y        298           0             0        WD   

  SaleCondition  
0        Normal  
1        Normal  

[2 rows x 54 columns]

In [ ]:
X_test.head(2)

MSSubClass MSZoning  LotFrontage  LotArea LotShape LandContour LotConfig  \
0          20       RH         80.0    11622      Reg         Lvl    Inside   
1          20       RL         81.0    14267      IR1         Lvl    Corner   

  LandSlope Neighborhood BldgType  ... GarageYrBlt  GarageFinish  GarageCars  \
0       Gtl        NAmes     1Fam  ...      1961.0           Unf         1.0   
1       Gtl        NAmes     1Fam  ...      1958.0           Unf         1.0   

   GarageArea  PavedDrive WoodDeckSF OpenPorchSF EnclosedPorch  SaleType  \
0       730.0           Y        140           0             0        WD   
1       312.0           Y        393          36             0        WD   

  SaleCondition  
0        Normal  
1        Normal  

[2 rows x 54 columns]

In [ ]:
# Apply preprocessing to the training data
X_train_transformed = preprocessor.fit_transform(X_train)
# Train the regressor on the transformed training data
rf_regressor.fit(X_train_transformed, y_train)

print("Preprocessing and model training complete.")


Preprocessing and model training complete.


In [ ]:
# original_test = pd.read_csv("/content/drive/MyDrive/Data Science Project/Datasets/test.csv")
original_test=pd.read_csv("test_data.csv")

In [ ]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# Apply preprocessing to the test data
X_test_transformed = preprocessor.transform(X_test)

# Make predictions on the transformed test data
y_pred = rf_regressor.predict(X_test_transformed)

# Create a submission DataFrame
submission_df = pd.DataFrame({'Id': original_test['Id'], 'SalePrice': y_pred})

# Save the submission DataFrame to a CSV file
submission_df.to_csv('submission.csv', index=False)

print("Submission file 'submission.csv' created successfully!")

# Calculate and print evaluation metrics
# Since we don't have the true y_test values for this submission,
# we cannot calculate metrics like R2, MAE, and MSE on the test set directly.
# However, we can calculate these metrics on the training set to assess the model's performance on seen data.

# Make predictions on the training data
y_train_pred = rf_regressor.predict(X_train_transformed)

# Calculate evaluation metrics on the training data
r2 = r2_score(y_train, y_train_pred)
mae = mean_absolute_error(y_train, y_train_pred)
mse = mean_squared_error(y_train, y_train_pred)

print("\nEvaluation Metrics on Training Data")
print(f"R-squared: {r2:.4f}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Mean Squared Error (MSE): {mse:.2f}")

Submission file 'submission.csv' created successfully!

--- Evaluation Metrics on Training Data ---
R-squared: 0.9808
Mean Absolute Error (MAE): 6501.01
Mean Squared Error (MSE): 121384517.15


## Evaluate for Overfitting

Based on the results from the hyperparameter tuning in cell `d64fc4da` and the evaluation of the tuned model on the training data in cell `bf973c3b`:

*   **R-squared on Training Data (Tuned Model):** `r2_tuned` (from cell `bf973c3b`)
*   **Best Cross-Validation R-squared during tuning:** `random_search.best_score_` (from cell `d64fc4da`)

A significant difference between the R-squared on the training data and the cross-validation R-squared suggests overfitting.

**Analysis:**

*   The R-squared on the full training data with the tuned model is {r2_tuned:.4f}.
*   The best average R-squared from 5-fold cross-validation during `RandomizedSearchCV` was {random_search.best_score_:.4f}.

Since the R-squared on the training data ({r2_tuned:.4f}) is much higher than the cross-validation R-squared ({random_search.best_score_:.4f}), it indicates that the tuned model is likely overfitting to the training data. The model is performing exceptionally well on the data it has seen but is not generalizing as well to slightly different subsets of the training data (as seen in cross-validation).

**Performance on Test Data:**

While we cannot calculate standard metrics (R-squared, MAE, MSE) on the provided test data because it lacks the target variable (`SalePrice`), the `submission_tuned.csv` file contains the predictions made by the tuned model on this unseen data. The true performance of the model on the test data will be revealed when this submission file is evaluated against the actual `SalePrice` values on the competition or evaluation platform. The cross-validation score ({random_search.best_score_:.4f}) is a more realistic estimate of how the model might perform on truly unseen data compared to the R-squared on the training data.

**Reasoning**:
Train and evaluate a RandomForestRegressor model with the best hyperparameters found by the wider RandomizedSearchCV search on the training data, and then make predictions on the test data.

# Task
Explain how to improve the model using hyperparameter tuning.

## Identify hyperparameters to tune

### Subtask:
Determine which hyperparameters of the `RandomForestRegressor` are most likely to impact performance.


**Reasoning**:
Identify the key hyperparameters of RandomForestRegressor that are likely to impact performance based on the scikit-learn documentation.



In [ ]:
# Key hyperparameters for RandomForestRegressor that are likely to impact performance include:
# n_estimators: The number of trees in the forest. Increasing this generally improves performance
#               but also increases computation time.
# max_depth: The maximum depth of the trees. Controls overfitting; deeper trees can capture
#            more complex relationships but are more prone to overfitting.
# min_samples_split: The minimum number of samples required to split an internal node.
#                    Controls overfitting; higher values prevent splitting on small groups.
# min_samples_leaf: The minimum number of samples required to be at a leaf node.
#                   Controls overfitting; higher values prevent creating leaves with very few samples.
# max_features: The number of features to consider when looking for the best split.
#               Controls the randomness of the splits and can help decorrelate the trees.
# bootstrap: Whether bootstrap samples are used when building trees. Can impact variance.

print("Identified hyperparameters for tuning:")
print("- n_estimators: Number of trees in the forest.")
print("- max_depth: Maximum depth of individual trees.")
print("- min_samples_split: Minimum number of samples required to split an internal node.")
print("- min_samples_leaf: Minimum number of samples required to be at a leaf node.")
print("- max_features: Number of features to consider for the best split.")
print("- bootstrap: Whether bootstrap samples are used.")

Identified hyperparameters for tuning:
- n_estimators: Number of trees in the forest.
- max_depth: Maximum depth of individual trees.
- min_samples_split: Minimum number of samples required to split an internal node.
- min_samples_leaf: Minimum number of samples required to be at a leaf node.
- max_features: Number of features to consider for the best split.
- bootstrap: Whether bootstrap samples are used.


## Define a search space

### Subtask:
Define a search space for the selected hyperparameters of the RandomForestRegressor.


**Reasoning**:
Create a dictionary defining the search space for the selected hyperparameters of the RandomForestRegressor model.



In [ ]:
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['sqrt', 'log2'],
    'bootstrap': [True, False]
}

print(param_grid)

{'n_estimators': [100, 200, 300], 'max_depth': [None, 10, 20, 30], 'min_samples_split': [2, 5, 10], 'min_samples_leaf': [1, 2, 4], 'max_features': ['sqrt', 'log2'], 'bootstrap': [True, False]}


## Choose a search method

### Subtask:
Select a strategy for searching the hyperparameter space (e.g., Grid Search, Random Search).


**Reasoning**:
Given the size of the parameter grid and the computational cost of training RandomForestRegressor, RandomizedSearchCV is a more efficient approach than GridSearchCV for exploring the hyperparameter space.



In [ ]:
from sklearn.model_selection import RandomizedSearchCV

# The number of iterations for RandomizedSearchCV can be adjusted based on computational resources
# A higher number of iterations explores more of the search space.
n_iter_search = 100 # We will use 100 iterations

# Create the RandomizedSearchCV object
random_search = RandomizedSearchCV(rf_regressor, param_grid, n_iter=n_iter_search, cv=5, random_state=42, n_jobs=-1)

print(f"Selected RandomizedSearchCV for hyperparameter tuning with {n_iter_search} iterations.")

Selected RandomizedSearchCV for hyperparameter tuning with 100 iterations.


## Implement the tuning process

### Subtask:
Use RandomizedSearchCV to find the best hyperparameters for the RandomForestRegressor using the defined search space and training data.


**Reasoning**:
Fit the RandomizedSearchCV object to the transformed training data and print the best parameters and score.



**Reasoning**:
Train and evaluate a RandomForestRegressor model with the best hyperparameters found by RandomizedSearchCV on the training data, then evaluate its performance on the test data.



In [ ]:
# Retrieve the best hyperparameters
best_params = random_search.best_params_
print("Best hyperparameters:", best_params)

# Instantiate a new RandomForestRegressor model with the best hyperparameters
rf_regressor_tuned = RandomForestRegressor(**best_params, random_state=42)

# Train the tuned model on the transformed training data
rf_regressor_tuned.fit(X_train_transformed, y_train)
print("Tuned model trained successfully.")

# Apply preprocessing to the test data (already done in a previous cell, but explicitly calling transform here)
X_test_transformed = preprocessor.transform(X_test)
print("Test data transformed.")

# Make predictions on the transformed test data
y_pred_tuned = rf_regressor_tuned.predict(X_test_transformed)
print("Predictions on test data made.")

# Create a submission DataFrame
submission_df_tuned = pd.DataFrame({'Id': original_test['Id'], 'SalePrice': y_pred_tuned})

# Save the submission DataFrame to a CSV file
submission_df_tuned.to_csv('submission_tuned.csv', index=False)
print("Submission file 'submission_tuned.csv' created successfully!")

# Make predictions on the training data using the tuned model
y_train_pred_tuned = rf_regressor_tuned.predict(X_train_transformed)

# Calculate and print evaluation metrics on the training data
r2_tuned = r2_score(y_train, y_train_pred_tuned)
mae_tuned = mean_absolute_error(y_train, y_train_pred_tuned)
mse_tuned = mean_squared_error(y_train, y_train_pred_tuned)

print("\n--- Evaluation Metrics on Training Data (Tuned Model) ---")
print(f"R-squared: {r2_tuned:.4f}")
print(f"Mean Absolute Error (MAE): {mae_tuned:.2f}")
print(f"Mean Squared Error (MSE): {mse_tuned:.2f}")

Best hyperparameters: {'n_estimators': 200, 'min_samples_split': 2, 'min_samples_leaf': 1, 'max_features': 'sqrt', 'max_depth': 20, 'bootstrap': False}
Tuned model trained successfully.
Test data transformed.
Predictions on test data made.
Submission file 'submission_tuned.csv' created successfully!

--- Evaluation Metrics on Training Data (Tuned Model) ---
R-squared: 1.0000
Mean Absolute Error (MAE): 243.36
Mean Squared Error (MSE): 196821.09


## Summary:

### Data Analysis Key Findings

* The key hyperparameters identified for tuning the `RandomForestRegressor` include `n_estimators`, `max_depth`, `min_samples_split`, `min_samples_leaf`, `max_features`, and `bootstrap`.
* A search space for these hyperparameters was defined using a dictionary `param_grid`.
* `RandomizedSearchCV` with 100 iterations was selected as the hyperparameter tuning method, using 5-fold cross-validation to assess performance and help mitigate overfitting during the tuning process.
* The best hyperparameters found by `RandomizedSearchCV` were: `{'n_estimators': 200, 'min_samples_split': 2, 'min_samples_leaf': 1, 'max_features': 'sqrt', 'max_depth': 20, 'bootstrap': False}`.
* The best R-squared score obtained during the `RandomizedSearchCV` (cross-validation score) was approximately 0.873.
* The tuned model achieved an R-squared of 1.0000, a Mean Absolute Error (MAE) of 243.36, and a Mean Squared Error (MSE) of 196821.09 on the training data.

### Insights or Next Steps

* The perfect R-squared score on the training data (1.0000) compared to the cross-validation R-squared (0.873) strongly suggests that the tuned model is significantly overfitting to the training data. While hyperparameter tuning with cross-validation helps, additional measures may be needed to improve generalization.
* To improve the model's performance on unseen data and reduce overfitting further, you could consider:
    * **Further Hyperparameter Tuning:** Explore a wider range or denser grid of hyperparameters around the best found values.
    * **Regularization Techniques:** While Random Forests are less prone to traditional regularization, techniques like limiting the number of features considered at each split (`max_features`) or increasing `min_samples_leaf` and `min_samples_split` can help.
    * **Ensemble Methods:** Experiment with other ensemble techniques like Gradient Boosting (e.g., XGBoost, LightGBM) which often provide strong performance and have built-in regularization parameters.
    * **Feature Engineering:** Create new features or modify existing ones to better capture underlying patterns and potentially reduce the complexity needed by the model.
    * **Collect More Data:** If feasible, increasing the size and diversity of the training data can help the model generalize better.
* The `submission_tuned.csv` file contains the predictions made by the tuned model on the unseen test data. The true performance of the model on the test data will be revealed when this submission file is evaluated against the actual `SalePrice` values on the competition or evaluation platform. The cross-validation score (0.873) is a more realistic estimate of how the model might perform on truly unseen data compared to the R-squared on the training data.